lib/hydra/works/services/generic_file/add_file.rb

```diff
@@ -1,31 +1,58 @@
 module Hydra::Works
   class AddFileToGenericFile
 
-    # Adds the original file to the work
+    # Adds a file to the work
 
-    def self.call(object, path, type, replace=false)
+    def self.call(object, path, type, update_existing=true)
       raise ArgumentError, "supplied object must be a generic file" unless Hydra::Works.generic_file?(object)
       raise ArgumentError, "supplied path must be a string" unless path.is_a?(String)
       raise ArgumentError, "supplied path to file does not exist" unless ::File.exists?(path)
 
-      if replace
-        current_file = object.filter_files_by_type(self.type_to_uri(type)).first
+      if type.instance_of? Symbol
+        raise ArgumentError, "you're attempting to add a file to a generic_file using '#{type}' association but the generic_file does not have an association called '#{type}''" unless object_has_association?(object, type)
+        if update_existing
+          current_file = object.send(type)
+        end
+        unless current_file
```

Comment:
`update_existing` just says "if there's something already there, update it rather than adding a new element", so there are plenty of cases where `update_existing` is set to true but object.send(type) returns nil.